src/components/Button/index.jsx

```diff
@@ -150,6 +152,50 @@ Button.propTypes = {
   Link: PropTypes.func,
   /** Extend click radius of button to nearest relative parent */
   isBreakoutLink: PropTypes.bool,
+  /** Allows specifying a margin class value. all--y, top--xl, etc */
+  margin(props, propName) {
+    // This will be split out into a reusable function the moment we
+    // would like to use this elsewhere.
+    const marginValue = props[propName];
+
+    const directions = {
+      all: true,
+      left: true,
+      right: true,
+      top: true,
+      bottom: true,
+      x: true,
+      y: true,
+    };
+
+    const amounts = {
+      xs: true,
+      s: true,
+      m: true,
+      l: true,
+      xl: true,
+    };
+
+    if (marginValue) {
+      const split = marginValue.split('--');
+      // If the first part of the split class is not a direction, it's an invalid value.
+      if (!directions[split[0]]) {
+        return new Error(`Invalid margin (no valid direction detected)`);
+      }
+      // If there happens to be an amount (which is optional)
+      if (split[1]) {
+        // If a standard amount isn't detected, the class is invalid
+        if (!amounts[split[1]]) {
+          return new Error(`Invalid margin value (no valid value detected)`);
```

Comment:
In CSS classes, the amount is optional - we always default to the `m` value.
Should we also allow this in the component?

```
<Button margin="x" />
```

Reply:
yep that's allowed. `if (!amounts[split[1]]) {` is only run if an amount exists (`split[1]`). If it doesn't , it just falls through.

Reply:
I broke these out into variables with appropriate names so that the code is easer to understand.